File: codigos/Time.java
package codigos;

public class Time {

    private String nome;
    private String tecnico;
    private String estado;

    private int vitorias = 0;
    private int empates = 0;
    private int derrotas = 0;
    private int golsPro = 0;
    private int golsContra = 0;

    private int vitoriasMandante = 0;
    private int vitoriasVisitante = 0;
    private int partidasJogadas = 0;

    // --------- Construtor ---------
    public Time(String nome, String tecnico, String estado) {
        this.nome = nome;
        this.tecnico = tecnico;
        this.estado = estado;
    }

    // --------- Atualiza estatísticas ---------
    public void atualizarEstatisticas(Partida p) {

        if (p.getMandante().equals(nome)) {

            golsPro += p.getPlacarMandante();
            golsContra += p.getPlacarVisitante();
            partidasJogadas++;

            if (p.getVencedor().equals(nome)) {
                vitorias++;
                vitoriasMandante++;
            } else if (p.getVencedor().equals("Empate")) {
                empates++;
            } else {
                derrotas++;
            }

        } else if (p.getVisitante().equals(nome)) {

            golsPro += p.getPlacarVisitante();
            golsContra += p.getPlacarMandante();
            partidasJogadas++;

            if (p.getVencedor().equals(nome)) {
                vitorias++;
                vitoriasVisitante++;
            } else if (p.getVencedor().equals("Empate")) {
                empates++;
            } else {
                derrotas++;
            }
        }
    }

    // --------- Getters ---------
    public int getVitorias() {
        return vitorias;
    }

    public int getEmpates() {
        return empates;
    }

    public int getDerrotas() {
        return derrotas;
    }

    public int getGolsPro() {
        return golsPro;
    }

    public int getGolsContra() {
        return golsContra;
    }

    public int getVitoriasMandante() {
        return vitoriasMandante;
    }

    public int getVitoriasVisitante() {
        return vitoriasVisitante;
    }

    public int getPartidasJogadas() {
        return partidasJogadas;
    }

    public String getNome() {
        return nome;
    }

    public String getTecnico() {
        return tecnico;
    }

    public String getEstado() {
        return estado;
    }

    // --------- Calculados ---------
    public int getPontuacao() {
        return vitorias * 3 + empates;
    }

    public int getSaldoGols() {
        return golsPro - golsContra;
    }

    public double getMediaGolsSofridos() {
        return partidasJogadas > 0 ? (double) golsContra / partidasJogadas : 0.0;
    }

    // --------- Representação textual ---------
    @Override
    public String toString() {
        return nome + " (" + estado + ")" +
               " - Pontos: " + getPontuacao() +
               ", Jogos: " + partidasJogadas +
               " (V: " + vitorias + ", E: " + empates + ", D: " + derrotas + ")" +
               ", Gols: " + golsPro + ":" + golsContra +
               ", Saldo: " + getSaldoGols() +
               ", Vitórias Casa: " + vitoriasMandante +
               ", Vitórias Fora: " + vitoriasVisitante +
               ", Média Gols Sofridos: " + String.format("%.2f", getMediaGolsSofridos());
    }
}

File: Python/codigosPython/Time.py
from .Partida import Partida

class Time:
    def __init__(self, nome: str, tecnico: str, estado: str):
        self._nome = nome
        self._tecnico = tecnico
        self._estado = estado

        self._vitorias = 0
        self._empates = 0
        self._derrotas = 0
        self._golsPro = 0
        self._golsContra = 0

        self._vitoriasMandante = 0
        self._vitoriasVisitante = 0
        self._partidasJogadas = 0

    # Getters
    @property
    def nome(self) -> str:
        return self._nome

    @property
    def tecnico(self) -> str:
        return self._tecnico

    @property
    def estado(self) -> str:
        return self._estado

    @property
    def vitorias(self) -> int:
        return self._vitorias

    @property
    def empates(self) -> int:
        return self._empates

    @property
    def derrotas(self) -> int:
        return self._derrotas

    @property
    def golsPro(self) -> int:
        return self._golsPro

    @property
    def golsContra(self) -> int:
        return self._golsContra

    @property
    def vitoriasMandante(self) -> int:
        return self._vitoriasMandante

    @property
    def vitoriasVisitante(self) -> int:
        return self._vitoriasVisitante

    @property
    def partidasJogadas(self) -> int:
        return self._partidasJogadas

    def atualizarEstatisticas(self, p: Partida):
        if p.mandante == self._nome:
            self._golsPro += p.placarMandante
            self._golsContra += p.placarVisitante
            self._partidasJogadas += 1

            if p.vencedor == self._nome:
                self._vitorias += 1
                self._vitoriasMandante += 1
            elif p.vencedor == "Empate":
                self._empates += 1
            else:
                self._derrotas += 1

        elif p.visitante == self._nome:
            self._golsPro += p.placarVisitante
            self._golsContra += p.placarMandante
            self._partidasJogadas += 1

            if p.vencedor == self._nome:
                self._vitorias += 1
                self._vitoriasVisitante += 1
            elif p.vencedor == "Empate":
                self._empates += 1
            else:
                self._derrotas += 1

    def getPontuacao(self) -> int:
        return self._vitorias * 3 + self._empates

    def getSaldoGols(self) -> int:
        return self._golsPro - self._golsContra

    def getMediaGolsSofridos(self) -> float:
        return self._golsContra / self._partidasJogadas if self._partidasJogadas > 0 else 0.0

    def __str__(self):
        return (
            f"{self._nome} - Pontos: {self.getPontuacao()}, "
            f"Gols: {self._golsPro}:{self._golsContra}, "
            f"Saldo: {self.getSaldoGols()}, "
            f"Vitórias Casa: {self._vitoriasMandante}, "
            f"Vitórias Fora: {self._vitoriasVisitante}, "
            f"Média Gols Sofridos: {self.getMediaGolsSofridos():.2f}"
        )
